### DISCUSSION TITLE
Table locking in orm?

### DISCUSSION THREAD

**Discussion by csm10495** [sqlalchemy/sqlalchemy#9723] [Ideas]
Hey folks,

In MySQL there is a lock table feature: https://dev.mysql.com/doc/refman/8.0/en/lock-tables.html

I can't really speak to other dialects but it would be cool if there was a way via orm to lock an entire table for either reads and/or writes.

Generally its not great to lock a whole table, but sometimes it can be necessary.

I picture having a contextmanager to have a table lock and ultimately release it.

Could this be supported? (If it already exists without falling back to raw sql let me know).

Thanks!

**Comment by zzzeek** [comment]
You can emit these statements directly:

```py
with engine.begin() as conn:
    conn.execute(text("LOCK TABLES ..."))
```

these are totally DB-specific commands so there's no reason to complicate it.  you can certainly make your own context manager:


```
import contextlib

@contextlib.contextmanger
def lock_tables(connection, tablenames):
    connection.execute(text("LOCK TABLES ...."))
    try:
        yield
    finally:
        # or whatever the command is, paraphrasing here
        connection.execute(text("UNLOCK TABLES ...."))

```

**Reply by csm10495** [reply]
Yeah this is basically what I did locally. I just wonder if there is value in trying to abstract it at a higher level across db types.

**Reply by zzzeek** [reply]
IMO it's not because other databases don't necessarily have this concept available, or it's not commonly used for other backends.   looks like PostgreSQL and Oracle have it also but I've never heard of this being used for those DBs.  That is this "feels" a bit like a "MySQL cultural" thing.

Note that the normal way to "lock" things is to use SELECT..FOR UPDATE, which we support everywhere.   Using this form might be a counterargument to your assertion that "lock table is sometimes necessary", since SELECT..FOR UPDATE is a more flexible way to lock rows.

**Reply by csm10495** [reply]
The problem is in my case I need to ensure the same row isn't inserted twice in a race condition.. unfortunately the underlying db replication platform doesn't support adding additional unique keys.

The only other option was to change the PK, but we couldn't do that since parts of it can be null.

Generally I think locking a whole table for an insert is crappy, but not being able to use unique keys in our platform led to this.

Even if not a great solution most of the time, is there not merit in abstracting the different dialects locks here?

**Reply by zzzeek** [reply]
I would not lock the table for that.   I would do that operation optimistically:

```py
try:
    with conn.begin_nested():  # uses SAVEPOINT
        conn.execute(insert_stmt)
except IntegrityError:
    print("row already inserted")
```

that way the rest of the world can keep communicating with that table.

**Reply by csm10495** [reply]
The problem is that the table is fine with the duplicate rows. (I'm probably using the term duplicate wrong but certain columns match in this case. Unfortunately the PK is different.)

So for us a duplicate means a certain combo of columns matches in multiple rows. Unfortunately the PK is different in both cases.

Wouldn't that only work if the full row was the same (so same insert statement) and the whole row was a PK?

**Reply by zzzeek** [reply]
OK so you have two unconstrained rows and you want to dedupe on an unconstrained column.     So yes, there's no way for the database to give you any atomicity on that and you have to lock the whole table.

But overall, this is not a situation I'd recommend people getting into and I'd typically guess "MySQL" if I heard this use case without knowing what DB it was, since MySQL for many years was an "unconstrained" kind of DB with MyISAM and all that.   This is not the kind of thing you see in Oracle / PG applications.

Basically LOCK TABLE is 99% of the time an antipattern, apps that use it are im going to say 100% of the time not designed to be DB agnostic anyway because as we see, using it here is at the end of a line of MySQL-specific things going on, and someone that really needs to use it can just emit the SQL.  that is I dont think this should be advertised as a top level API feature in the main library.

**Reply by csm10495** [reply]
Fair enough. Thanks still.

.. I wish my case would let us just add a unique index since that would make the problem go away.. but alas.
